File: platforms/ios/Lists/Core/Queries/ItemSort.swift
import Foundation

extension Array where Element == Item {
    /// Re-sorts items according to a user-chosen `SortMode`. `.manual`
    /// preserves the existing order; the other modes apply the named
    /// comparator (with a stable title tiebreaker for `.priority`).
    func sortedBy(
        _ mode: ListViewPreferences.SortMode,
        direction: ListViewPreferences.SortDirection = .ascending
    ) -> [Item] {
        let ascending: [Item]
        switch mode {
        case .manual:
            // Manual drag order — direction is ignored.
            return sorted { $0.sortIndex < $1.sortIndex }
        case .due:
            ascending = sorted { ($0.due ?? .distantFuture) < ($1.due ?? .distantFuture) }
        case .alphabetical:
            ascending = sorted {
                $0.title.localizedCaseInsensitiveCompare($1.title) == .orderedAscending
            }
        case .created:
            ascending = sorted { $0.createdAt < $1.createdAt }
        case .priority:
            ascending = sorted { lhs, rhs in
                let l = Self.priorityRank(lhs.priority)
                let r = Self.priorityRank(rhs.priority)
                if l != r { return l < r }
                return lhs.title.localizedCaseInsensitiveCompare(rhs.title) == .orderedAscending
            }
        }
        return direction == .descending ? ascending.reversed() : ascending
    }

    private static func priorityRank(_ p: Item.Priority) -> Int {
        switch p {
        case .high:   return 0
        case .medium: return 1
        case .low:    return 2
        case .none:   return 3
        }
    }
}

File: platforms/android/core/src/main/kotlin/io/github/saxonbobart/lists/core/query/ItemSort.kt
package io.github.saxonbobart.lists.core.query

import io.github.saxonbobart.lists.core.model.Item
import io.github.saxonbobart.lists.core.model.Priority
import java.time.Instant

enum class SortMode { MANUAL, DUE, ALPHABETICAL, CREATED, PRIORITY }

enum class SortDirection { ASCENDING, DESCENDING }

/**
 * Re-sorts items by a user-chosen [SortMode]. MANUAL preserves drag order
 * (`sortIndex`, direction ignored); the others apply the named comparator
 * with a stable title tiebreaker for PRIORITY.
 */
fun List<Item>.sortedByMode(
    mode: SortMode,
    direction: SortDirection = SortDirection.ASCENDING,
): List<Item> {
    val ascending = when (mode) {
        SortMode.MANUAL -> return sortedBy { it.sortIndex }
        SortMode.DUE -> sortedBy { it.due ?: Instant.MAX }
        SortMode.ALPHABETICAL -> sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER) { it.title })
        SortMode.CREATED -> sortedBy { it.createdAt }
        SortMode.PRIORITY -> sortedWith(
            compareBy<Item> { priorityRank(it.priority) }
                .thenBy(String.CASE_INSENSITIVE_ORDER) { it.title },
        )
    }
    return if (direction == SortDirection.DESCENDING) ascending.reversed() else ascending
}

private fun priorityRank(p: Priority): Int = when (p) {
    Priority.HIGH -> 0
    Priority.MEDIUM -> 1
    Priority.LOW -> 2
    Priority.NONE -> 3
}
